databricks/koalas/frame.py

```diff
@@ -6810,6 +6810,213 @@ def filter(self, items=None, like=None, regex=None, axis=None):
+                    raise KeyError('Index include label which is not in the `mapper`.')
+                else:
+                    return DataFrame(self._internal)
+
+            type_set = set(map(lambda x: type(x), mapper.values()))
+            if len(type_set) > 1:
+                raise ValueError("Mapper dict should have the same value type.")
+            spark_return_type = as_spark_type(list(type_set)[0])
+
+            def mapper_fn(x):
+                if x in mapper:
+                    return mapper[x]
+                else:
+                    if errors == 'raise':
+                        raise KeyError('Index include value which is not in the `mapper`')
+                    return x
+        elif callable(mapper):
+            spark_return_type = _infer_return_type(mapper).tpe
+
+            def mapper_fn(x):
+                return mapper(x)
+        else:
+            raise ValueError("`mapper` or `index` or `columns` should be "
+                             "either dict-like or function type.")
+
+        if is_index_mapper:
+            # rename index labels, if `level` is None, rename all index columns, otherwise only
+            # rename the corresponding level index.
+            # implement this by transform the underlying spark dataframe,
+            # Example:
+            # suppose the kdf index column in underlying spark dataframe is "index_0", "index_1",
+            # if rename level 0 index labels, will do:
+            #   ``kdf._sdf.withColumn("index_0", mapper_fn_udf(col("index_0"))``
+            # if rename all index labels (`level` is None), then will do:
+            #   ```
+            #   kdf._sdf.withColumn("index_0", mapper_fn_udf(col("index_0"))
+            #           .withColumn("index_1", mapper_fn_udf(col("index_1"))
+            #   ```
+
+            index_columns = self._internal.index_columns
+            num_indices = len(index_columns)
+            if level:
+                if level < 0 or level >= num_indices:
+                    raise ValueError("level should be an integer between [0, num_indices)")
+
+            def gen_new_index_column(level):
+                index_col_name = index_columns[level]
+
+                index_mapper_udf = pandas_udf(lambda s: s.map(mapper_fn),
+                                              returnType=spark_return_type)
+                return index_mapper_udf(col(index_col_name))
+
+            sdf = self._sdf
+            if level is None:
+                for i in range(num_indices):
+                    sdf = sdf.withColumn(index_columns[i], gen_new_index_column(i))
+            else:
+                sdf = sdf.withColumn(index_columns[level], gen_new_index_column(level))
+
+            internal = self._internal.copy(sdf=sdf)
+        else:
+            # rename column name.
+            # we only need to modify the `kdf._internal._column_index
+            # replace each column name by mapper dict or mapper function.
```

Comment:
The comment is right in the future, but unfortunately so far we don't fully support it.
We still need to transform the underlying Spark DataFrame to have the same column name for single-index columns, or string of the tuple for multi-index columns.